estimagic/dashboard/convergence_plot.py

```diff
@@ -0,0 +1,157 @@
+"""Functions for creating and styling the convergence plot."""
+import random
+from functools import partial
+
+import bokeh.palettes
+from bokeh.models import ColumnDataSource
+from bokeh.plotting import figure
+from tornado import gen
+
+
+def setup_convergence_plot(param_df, initial_fitness):
+    """
+    Setup the convergence plot for later updating.
+
+    This function is called in _setup_dashboard.
+
+    Args:
+        param_df (pandas DataFrame):
+            DataFrame with the initial parameter values, constraints etc.
+        initial_fitness (pd.Series):
+            criterion function evaluated at the initial parameters
+
+    """
+    fitness_plot, fitness_data = _fitness_plot(initial_fitness=initial_fitness)
+    param_plots, param_data = _parameter_plots(param_df=param_df)
+    plots = [fitness_plot] + param_plots
+    datasets = [fitness_data] + param_data
+    return plots, datasets
+
+
+def update_convergence_plot(doc, queue, datasets):
+    """
+    Check for new param values and update the plot.
+
+    This function is called in a never ending while loop in _update_dashboard.
+
+    Args:
+        doc (bokeh Document):
+            document instance where the Dashboard will be stored.
+            Note this must stay the first argument for the bokeh FunctionHandler
+            to work properly.
+
+        queue (Queue):
+            queue to which originally the parameters DataFrame is supplied and to which
+            the updated parameter Series will be supplied later.
+
+        datasets (list):
+            list of ColumnDataSource storing earlier parameter iterations.
+
+    """
+    fitness_data, param_data = datasets
+
+    if queue.qsize() > 0:
+        new_params, fitness = queue.get()
+
+        doc.add_next_tick_callback(
+            partial(_update_convergence_plot, data=param_data, new_values=new_params)
+        )
+        doc.add_next_tick_callback(
+            partial(_update_convergence_plot, data=fitness_data, new_values=fitness)
+        )
+
+
+def _fitness_plot(initial_fitness):
+    fitness_data = ColumnDataSource(
+        data=_convert_sr_for_cds(sr=initial_fitness, iteration=0)
+    )
+    fitness_p = _wide_figure(title="Fitness")
+    fitness_p.line(
+        source=fitness_data,
+        x="iteration",
+        y="fitness",
+        line_width=1,
+        name="fitness",
+        color="firebrick",
+    )
+
+    return fitness_p, fitness_data
+
+
+def _parameter_plots(param_df):
+    """
+    Create the plots that show the convergence of (groups of) parameters.
+
+    ToDo: split up convergence plot depending on column in param_df.
+    ToDo: plot upper and lower bounds.
+    ToDo: only plot parameters that are not fixed.
+
+    Args:
+        param_df (pandas DataFrame):
+            DataFrame with the initial parameter values, constraints etc.
+
+
+    """
+
+    assert "iteration" not in param_df.index, (
```

Comment:
If at all, this error handling should be [here](https://github.com/janosg/estimagic/blob/0a57468954e20de59af420f596012a0b62622533/estimagic/optimization/optimize.py#L141)

But even better would be not to have this restriction. Is 'iteration' displayed somewher? Otherwise we could use a long and cryptic name that is essentially guaranteed to be unique.

Reply:
At the moment it's not displayed anywhere but I am planning to put it as the x axis label later. I can hard code that though and avoid needing a "pretty" column name.